bundles/org.openhab.binding.vdr/src/main/java/org/openhab/binding/vdr/internal/svdrp/SVDRPClientImpl.java

```diff
@@ -0,0 +1,409 @@
+/**
+ * Copyright (c) 2010-2021 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.vdr.internal.svdrp;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link SVDRPClientImpl} encapsulates all calls to the SVDRP interface of a VDR
+ *
+ * @author Matthias Klocke - Initial contribution
+ */
+@NonNullByDefault
+public class SVDRPClientImpl implements SVDRPClient {
+
+    private final Logger logger = LoggerFactory.getLogger(SVDRPClientImpl.class);
+
+    private String host;
+    private int port = 6419;
+    private String charset = "UTF-8";
+    private String version = "";
+
+    private final String welcomeMessagePattern = "([0-9]{3})([ -])(.*)";
+    private Pattern patternMessage = Pattern.compile(welcomeMessagePattern);
+
+    private static int timeout = 3000;
+
+    private @Nullable Socket socket = null;
+    private @Nullable BufferedWriter out = null;
+    private @Nullable BufferedReader in = null;
+
+    public SVDRPClientImpl(String host, int port) {
+        super();
+        this.host = host;
+        this.port = port;
+    }
+
+    public SVDRPClientImpl(String host, int port, String charset) {
+        super();
+        this.host = host;
+        this.port = port;
+        this.charset = charset;
+    }
+
+    @Override
+    protected void finalize() {
+        try {
+            this.closeConnection();
```

Comment:
`finalize()` should return fast. Is it really necessary to close the connection when the garbage collector comes? Actually you should ensure that it is always closed in the normal program flow.